core/types/subtyping.cc

```diff
@@ -1303,6 +1303,13 @@ bool Types::isSubTypeUnderConstraint(const GlobalState &gs, TypeConstraint &cons
                Types::isSubTypeUnderConstraint(gs, constr, t1, o2->right, mode);
     }
     if (a1 != nullptr) { // 4
+        auto *t2v = cast_type<TypeVar>(t2);
+        if (t2v && !constr.isSolved() && constr.hasLowerBound(t2v->sym) &&
+            ((const TypeConstraint &)constr).findLowerBound(t2v->sym) == Types::bottom()) {
```

Comment:
Why are we casting here?

Reply:
The non-`const` overload is selected here, but it is `private` and therefore not applicable.  I don't understand why the `const` overload can't be considered for resolution here, but it's probably because I have not sufficiently internalized the C++ overload rules.

Reply:
Another option for this case would be the following:
```
     if (a1 != nullptr) { // 4
         auto l = Types::isSubTypeUnderConstraint(gs, constr, a1->left, t2, mode);
         auto r = Types::isSubTypeUnderConstraint(gs, constr, a1->right, t2, mode);
         return l || r;
     }
```
The problem is that `Types::isSubTypeUnderConstraint` will learn new constraints as it processes types, so short-circuiting while processing the AndType in the LHS of the subtyping relation causes some constraints to be forgotten.

Reply:
We could do short-circuiting if `!isa_type<TypeVar>(t2)` to only take a small perf hit for the generic case.

Reply:
That seems reasonable to me. Would you mind running this on stripe's codebase, and seeing what sort of performance hit we would have without the special-case for TypeVar?

Reply:
I'm kind of thinking maybe we should just yoink these lines out of isSubtypeUnderConstraintSingle into  the top of isSubtypeUnderconstraint?

https://github.com/sorbet/sorbet/blob/a7b4be1fbe928c129b7cc02cdae161bab2f33f4f/core/types/subtyping.cc#L990-L996

Reply:
(we should test that this is correct and fast though, i don't claim for sure that it's either)

Reply:
I don't think changing case 4 in `isSubTypeUnderConstraint` to:

```c++
    if (a1 != nullptr) { // 4
        bool l = Types::isSubTypeUnderConstraint(gs, constr, a1->left, t2, mode);
        bool r = Types::isSubTypeUnderConstraint(gs, constr, a1->right, t2, mode);
        return l || r;
    }
```

works.  Running that change over the added test cases from the linked bug (which are added in this PR) gives:

```
test/testdata/infer/generic_all.rb:17: ERROR: Expected error of form:
17:     T.reveal_type(ab_list) # error: Revealed type: `T::Array[T.all(One::A, One::B)]`
    ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ error: Revealed type: `T::Array[T.all(One::A, One::B)]`
Found error:
17:     T.reveal_type(ab_list) # error: Revealed type: `T::Array[T.all(One::A, One::B)]`
        ^^^^^^^^^^^^^^^^^^^^^^ error: Revealed type: `T::Array[T.any(One::A, One::B)]`

test/testdata/infer/generic_all.rb:32: ERROR: Expected error of form:
32:     T.reveal_type(x.to_h) # error: Revealed type: `T::Hash[Symbol, T.all(Two::Interface, Two::Parent)]`
    ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ error: Revealed type: `T::Hash[Symbol, T.all(Two::Interface, Two::Parent)]`
Found error:
32:     T.reveal_type(x.to_h) # error: Revealed type: `T::Hash[Symbol, T.all(Two::Interface, Two::Parent)]`
        ^^^^^^^^^^^^^^^^^^^^^ error: Revealed type: `T::Hash[Symbol, T.any(Two::Interface, Two::Parent)]`

test/testdata/infer/generic_all.rb:63: ERROR: Found unexpected error:
63:       results
          ^^^^^^^ error: Returning value that does not conform to method result type
Note: If there is already an assertion for this error, then this is a duplicate error. Change the assertion to `# error-with-dupes: <error message>` if the duplicate is expected.
```

So we're not throwing away information, at least, but we're deriving `T.any` instead of `T.all`.

We are running into the asymmetry in `TypeConstraint::rememberIsSubtype`, and so we have to account for that somehow.